### PR TITLE
BlocksWpQuery wrapper for WP_Query

### PR DESCRIPTION
**Pull request by mikejolley**
As an alternative to #1056 which uses a custom hook to invalidate caches, this PR introduces a new wrapper for `WP_Query` to create a cache based on parsed query args in a standardized manner.

`BlocksWpQuery` is a class that wraps `WP_Query` and has 2 differences:

1. A constructor that doesn't run the query, just sets the query vars and parses them.
2. A new `get_cached_posts` method which attempts to pull results from a transient, falling back to `get_posts` if needed.

Closes #1056

### How to test the changes in this Pull Request:

1. Create a products by category block
2. View / Refresh, ensure results are set.
3. Manually check DB to ensure a transient with prefix `wc_blocks_query_` exists.

### CODE REVIEW

**Comment by strategio** [comment]
This looks good to me @mikejolley.

I had some troubles to move to the correct branch into the WC package but I manage to adapt my code and eventually test the change.

Just one comment, the variable `\WP_Query::query_vars_hash` is private but we manage to read/write it because `\WP_Query` has magic method.

For the record, here's my adjusted code:

```php
add_action( 'parse_query', [ $this, 'addCurrentLangToQueryVarsHash' ] );

public function addCurrentLangToQueryVarsHash( $query ) {
	if ( $query instanceof \Automattic\WooCommerce\Blocks\Utils\BlocksWpQuery ) {
		$query->query_vars_hash = md5( $query->query_vars_hash . $this->sitepress->get_current_language() );
	}
}
```

**Comment by mikejolley** [comment]
@strategio That code has missed the point. You said that WPML would inject query vars based on the current language? If that is the case there is zero need to modify the hash. The hash is built based on the `query_vars` `WP_Query` is using.

**Comment by strategio** [comment]
@mikejolley, in `\WP_Query::parse_query_vars`, the only hook is `parse_query` and it's called after `$this->query_vars_hash = md5( serialize( $this->query_vars ) );`.

And even though, as I mentioned in [this comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1056#issuecomment-545174209) (and perhaps I wasn't clear), we have filters in several places.

For instance, when we have some IDs (post / terms), we filter it `parse_query`. But this is not possible for archive queries for instance. So we also have some filters on the SQL clauses (JOIN/WHERE) in `\WP_Query`.

**Comment by nerrad** [comment]
@strategio I think we're having trouble grasping why things aren't working for you and it may be because we don't fully grasp what you are filtering for query results to be adjusted.  Would you mind summarizing the filters you are currently hooking into to adjust the queries and where in the WP loading process you execute the `add_filters` for those filters?  Ideally, if we had a simple proof of concept we could work with that will help us see your problem.

We're wanting to work with you here, but we're really reluctant to just wrap the existing hashing mechanism in a filter as it makes it fragile.

**Comment by strategio** [comment]
@nerrad, there are 2 issues with the current solution:

1. `BlocksWpQuery` will call the method `parse_query_vars` (which actually calls `parse_query`) in the constructor:

```php
public function parse_query( $query = '' ) {
	//...

	$this->query_vars_hash    = md5( serialize( $this->query_vars ) );
	$this->query_vars_changed = false;

	do_action_ref_array( 'parse_query', array( &$this ) );
}
```

As you can see, the only place where we can filter the query vars is in the hook `parse_query` which is located after the property assignment for `$this->query_vars_hash`.

So, even if my plugin changes the query vars in `parse_query` hook, it would not change `$this->query_vars_hash`. At least not automatically, that's why I need to do this from WPML.


2. In WPML we filter the query vars when it's possible. If we have some post IDs or term IDs, we can replace it with the ones in the correct language. However, sometimes `WP_Query` has no specific IDs. Think of the following query `new \WP_Query( [ 'post_type' => 'product' ] )` which will return the lastest products. WPML needs to filter the products by language and to do so we have filters on the SQL query built in `\WP_Query::get_posts` (i.e. `posts_where` and  `posts_join`).

The second issue will be covered if I can just add the current language to the cache args (pasting again an extract of my code):

```php
class Hooks {

	public function add_hooks() {
		add_action( 'parse_query', [ $this, 'addCurrentLangToQueryVarsHash' ] );
	}

	public function addCurrentLangToQueryVarsHash( $query ) {
		if ( $query instanceof \Automattic\WooCommerce\Blocks\Utils\BlocksWpQuery ) {
			$query->query_vars_hash = md5( $query->query_vars_hash . $this->sitepress->get_current_language() );
		}
	}
}

```

Note: This code is just for a compatibility plugin (WooCommerce Multilingual). The core logic to filter query vars and SQL queries by language is located in the main plugin (WPML).


As a conclusion, I would say that I can accommodate myself with `BlocksWpQuery`. But I think that it's overcomplicated, and less efficient:
- We always need to parse the query vars before to check the cache (which also means calling the hook `parse_query`).
- If a 3rd part plugin changes something in `parse_query` hook, it also needs to manually refresh `query_vars_hash` (this is not obvious).

**Comment by mikejolley** [comment]
I think there are 2 changes to make here in this case,

1) in this PR, create a custom hash after args are passed and filters are ran
2) in WPML, ensure the global you're using for languages is always injected into WP_Query so it is always reflected in the hash. Custom args won't break other queries, plus you can use this in your custom SQL callbacks to avoid using globals if you wanted. Then any hash based query in blocks or other plugins should in theory account for language.

**Comment by mikejolley** [comment]
Added inline docs to cover this.

File: src/Utils/BlocksWpQuery.php [hunk 1]
```diff
@@ -38,13 +38,24 @@ public function __construct( $query = '' ) {
...
 
 	/**
-	 * Cache cached posts, if a cache exists.
+	 * Get cached posts, if a cache exists.
+	 *
+	 * A hash is generated using the array of query_vars. If doing custom queries via filters such as posts_where
+	 * (where the SQL query is manipulated directly) you can still ensure there is a unique hash by injecting custom
+	 * query vars via the parse_query filter. For example:
+	 *
+	 *      add_filter( 'parse_query', function( $wp ) {
```
Comment (strategio) [inline src/Utils/BlocksWpQuery.php]: @mikejolley, I think you should rename `$wp` with `$query`.

**Comment by strategio** [comment]
Thanks @mikejolley, this is perfect for me, we will just add our extra argument to the query now.

File: src/Utils/BlocksWpQuery.php (hunk 1 above)
Comment (nerrad) [reply src/Utils/BlocksWpQuery.php]: Agreed.

**Pull request merged**